alonzo/impl/src/Cardano/Ledger/Alonzo/Tx.hs

```diff
@@ -261,7 +378,193 @@ indexedRdmrs ::
   ) =>
   Tx era ->
   ScriptPurpose (Crypto era) ->
-  Maybe (Data era)
-indexedRdmrs tx sp = Map.lookup policyid (txrdmrs . wits $ tx)
+  Maybe (Data era, ExUnits)
+indexedRdmrs tx sp = Map.lookup policyid (txrdmrs . txwits $ tx)
   where
     policyid = rdptr (body tx) sp
+
+-- ===============================================================
+-- From the specification, Figure 7 "Script Validation, cont."
+-- ===============================================================
+
+-- | valContext collects info from the Tx and the UTxO and translates it into
+--   A Data, which the Plutus language know how to interpret.
+valContext :: UTxO era -> Tx era -> ScriptPurpose (Crypto era) -> [Data era]
+valContext _utxo _tx _sp = []
+
+--TODO FIX THIS, when defined will always return singleton list
+-- see also: collectNNScriptInputs    where it is called
+
+-- TODO  Specification says CostMod, not CostModel
+runPLCScript ::
+  CostModel ->
+  AlonzoScript.Script era ->
+  [Data era] ->
+  ExUnits ->
+  (IsValidating, ExUnits)
+runPLCScript _cost _script _data _exunits = (IsValidating True, ExUnits 0 0) -- TODO FIX THIS
+
+-- ===============================================================
+-- From the specification, Figure 8 "Scripts and their Arguments"
+-- ===============================================================
+
+getData ::
+  forall era.
+  ( ToCBOR (Core.AuxiliaryData era),
+    ToCBOR (Core.Script era),
+    UsesTxOut era,
+    HasField "datahash" (Core.TxOut era) (Maybe (DataHash (Crypto era)))
+  ) =>
+  Tx era ->
+  UTxO era ->
+  ScriptPurpose (Crypto era) ->
+  [Data era]
+getData tx (UTxO m) sp = case sp of
+  Minting _policyid -> []
+  Rewarding _rewaccnt -> []
+  Certifying _dcert -> []
+  Spending txin ->
+    case Map.lookup txin m of
+      Nothing -> []
+      Just txout ->
+        case getField @"datahash" txout of
+          Nothing -> []
+          Just hash ->
+            case Map.lookup hash (witsData (wits tx)) of
+              Nothing -> []
+              Just d -> [d]
+
+-- IS IT REALLY THE CASE ONLY SPENDING PURPOSE HAS DATA? TODO
```

Comment:
@WhatisRT ?

Reply:
yes, all the other purposes are essentially state-less, so they don't have data

Reply:
fixed the comment